src/oc_report_buffer.erl

```diff
@@ -0,0 +1,114 @@
+%%%------------------------------------------------------------------------
+%% Copyright 2017, OpenCensus Authors
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc Buffer of trace spans to be reported.
+%% @end
+%%%-----------------------------------------------------------------------
+-module(oc_report_buffer).
+
+-behaviour(gen_server).
+
+-export([start_link/0,
+         store_span/1]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2]).
+
+-include("opencensus.hrl").
+
+-record(state, {reporter :: module(),
+                reporter_config :: #{},
+                send_interval_ms :: integer(),
+                timer_ref :: reference()}).
+
+-define(BUFFER_1, oc_report_buffer1).
+-define(BUFFER_2, oc_report_buffer2).
+-define(BUFFER_STATUS, oc_report_status).
+
+start_link() ->
+    maybe_init_ets(),
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec store_span(opencensus:span()) -> ok | {error, invalid_span}.
+store_span(Span=#span{}) ->
+    [{_, Buffer}] = ets:lookup(?BUFFER_STATUS, current_buffer),
+    ets:insert(Buffer, Span);
+store_span(_) ->
+    {error, invalid_span}.
+
+init(_Args) ->
+    SendInterval = application:get_env(opencensus, send_interval_ms, 500),
+    Reporter = application:get_env(opencensus, reporter, undefined),
+    ReporterConfig = Reporter:init(),
```

Comment:
It would be nice to pass arguments to init too, that way it can get configuration in a standardized way. (see the pid reporter who reads the environmental config instead of a general config that can be passed)

Reply:
Any suggestion on where to store the configuration, as in under what path?

```
{opencensus, [{reporter, {ReporterModule, ReporterArgs}]}.
```

```
{opencensus, [{reporter, ReporterModule},
                        {reporter_args, ReporterArgs}]}.
```

Reply:
I can also never decide where best to initially read in the application env. Sometimes I do it in the supervisor so we'd have `oc_report_buffer:start_link(Module, Args)`, with the idea that then tests or something can just call that directly instead of setting application env.

Reply:
Personally, I like the first variant as it prevents something like
```erlang
{opencensus, [
  {reporter_args, ReporterArgs}
 %% a lot of other stuff
 %% tuff
 %% uff
 %% ff
 %% f
 %%
{reporter, ReporterModule}]}.
```
it bundles things that belong together together, not split them out